### DISCUSSION TITLE
220M pre-aggregated native file, help finding the issue

### DISCUSSION THREAD

**Discussion by ooliver1** [bloomberg/memray#444] [Q&A]
This initial post is outdated, [this](https://github.com/bloomberg/memray/discussions/444#discussioncomment-6959864) is the current comment thread.

I have a 91M memray bin file i converted into html using flamegraph, which itself is 605M in size. How do you think I should process this?

Table is basically the same. I would like to know if there is a way to see if the browser is even attempting to process this data or not, since I see it is all in a script.

**Comment by ooliver1** [comment]
I suppose I could use this as my help discussion.

```
oliver in ~ λ memray tree memray.28.bin 

Allocation metadata
-------------------
Command line arguments: '/root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/memray/__main__.py run 
--aggregate -o wow/memray.28.bin -m vibr'
Peak memory size: 2.284GB
Number of allocations: 6325

Biggest 10 allocations:
-----------------------
📂 984.146MB (100.00 %) <ROOT>  
├── [[10 frames hidden in 6 file(s)]]
│   └── 📄 114.545MB (11.64 %) json  
│       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/aiohttp/client_reqrep.py:1120
├── [[2 frames hidden in 1 file(s)]]
│   └── 📂 268.538MB (27.29 %) __init__  
│       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/guild.py:285
│       ├── [[1 frames hidden in 1 file(s)]]
│       │   └── 📂 201.972MB (20.52 %) _sync  
│       │       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/guild.py:439
│       │       ├── [[1 frames hidden in 1 file(s)]]
│       │       │   └── 📂 117.000MB (11.89 %) _update  
│       │       │       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/channel.py:191
│       │       │       ├── 📄 65.000MB (6.60 %) _fill_overwrites  
│       │       │       │   /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/abc.py:444
│       │       │       └── 📄 52.000MB (5.28 %) _fill_overwrites  
│       │       │           /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/abc.py:445
│       │       └── 📄 42.972MB (4.37 %) _add_channel  
│       │           /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/guild.py:288
│       ├── [[2 frames hidden in 2 file(s)]]
│       │   └── 📄 34.000MB (3.45 %) _update  
│       │       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/role.py:237
│       └── 📄 32.566MB (3.31 %) _from_data  
│           /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/guild.py:409
├── [[1 frames hidden in 1 file(s)]]
│   └── 📄 32.001MB (3.25 %) data_received  
│       /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/aiohttp/client_proto.py:213
└── [[6 frames hidden in 6 file(s)]]
    └── 📄 24.004MB (2.44 %) received_message  
        /root/.cache/pypoetry/virtualenvs/music-TFcQMFAJ-py3.11/lib/python3.11/site-packages/nextcord/gateway.py:462
```

The guilds is fine and expected. The issue is, the resident memory size of the python process steadily increases from between 1 and 2GB all the way up until it runs out of memory. So I suppose this is the stack memory if the heap is happily doing nothing out of the ordinary, do you know how I can debug and deal with this?

**Reply by godlygeek** [reply]
> The issue is, the resident memory size of the python process steadily increases from between 1 and 2GB all the way up until it runs out of memory.

How much memory does it reach before it runs out of memory? 

> So I suppose this is the stack memory if the heap is happily doing nothing out of the ordinary

That seems unlikely to me. The Python interpreter really doesn't store anything large on the stack, and the call stack's depth is guaranteed to be fairly limited thanks to `sys.setrecursionlimit`. I guess it's possible that you've got a C extension that's putting large stuff on the stack, but that wouldn't be my guess. The increase in RSS might be due to something that Memray isn't tracking. It's possible that they're coming from some uncommon allocator that we're not hooking, but it seems more likely that they're something that is neither stack memory nor heap memory - like shared memory or loaded shared libraries, perhaps. If this is on Linux (or Windows + WSL) you might find something informative about what mappings are taking up the space in `/proc/$PID/maps` at the point where its memory has begun to rise higher than you expect.

**Reply by ooliver1** [reply]
> How much memory does it reach before it runs out of memory?

It is ran in two separate (almost identical, sharding for a Discord bot) processes, this example was only about 3GB (worried it would spike up overnight so I wanted to get the file early).

It has (from metrics) gone all the way to 6GB (when the other process was OOM killed).

![image](https://github.com/bloomberg/memray/assets/34910574/bb64d6ac-efc5-45ce-9d22-ea30517162cd)

This is what the grafana+prometheus metrics looks like for resident memory, it just steadily rises over several days, and this is what it looks like right now:

![image](https://github.com/bloomberg/memray/assets/34910574/ca4db3e2-4a95-46ed-928b-98e5fec698a4)


I'll take your advice of running with the native flag, hopefully it doesn't hurt performance so bad this time with pre-aggregated stats. Hopefully I'll see you in a few days with a bit of a happier story in #226 or at least some better debuggability :heart:

**Comment by ooliver1** [comment]
![image](https://github.com/bloomberg/memray/assets/34910574/56c0f989-c842-4132-8916-64a197cd7d62)

Here is my summary, `???` might be a C-extension, right? If it is then I suppose I could run with the native flags, but that had a significant performance impact last time I tried.

**Reply by godlygeek** [reply]
That's telling us that, at the point where that allocation happened, there were no Python frames on the stack. It could be a C extension causing that: if an extension spawns its own thread using the native C/C++/Rust threading APIs, without going through the Python `threading` module, that thread wouldn't start with any Python frames on the stack, and so we wouldn't be able to tell you anything about _where_ allocations it performs happened, other than their thread ID: there's no Python stack, and you didn't run with `--native` so we didn't collect a native stack.  If I'm right that it's a native thread causing this, and you need to learn something about where it's performing those allocations, I think you'll need to bite the bullet and use `--native`.

**Reply by ooliver1** [reply]
I had some issues trying to run `python3.11-dbg` with `python:3.11-slim-bookworm`

`Dockerfile`
```dockerfile
FROM python:3.11-slim-bookworm

WORKDIR /bot

RUN apt-get update && apt-get install git netcat-openbsd iproute2 debuginfod python3.11-dbg -yqq && rm -rf /var/lib/apt/lists/*

ENV DEBUGINFOD_URLS="https://debuginfod.elfutils.org/"
ENV POETRY_VIRTUALENVS_CREATE="false"
ENV PYTHONPATH="/usr/local/lib/python311.zip:/usr/local/lib/python3.11:/usr/local/lib/python3.11/lib-dynload:/usr/local/lib/python3.11/site-packages"
RUN pip install poetry

COPY vibr/pyproject.toml vibr/poetry.lock ./

RUN poetry install --no-root --only main

COPY . .
RUN mkdir wow

CMD ["sh", "vibr/entrypoint.sh"]
```

command: `python3.11-dbg -m memray run --aggregate --native -m vibr`

error:

```
music-bot-1            | Memray WARNING: Correcting symbol for malloc from 0x41f610 to 0x7febc3b99770
music-bot-1            | Memray WARNING: Correcting symbol for free from 0x41faa0 to 0x7febc3b99d30
music-bot-1            | Traceback (most recent call last):
music-bot-1            |   File "/usr/local/lib/python3.11/runpy.py", line 198, in _run_module_as_main
music-bot-1            |     return _run_code(code, main_globals, None,
music-bot-1            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
music-bot-1            |   File "/usr/local/lib/python3.11/runpy.py", line 88, in _run_code
music-bot-1            |     exec(code, run_globals)
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/memray/__main__.py", line 6, in <module>
music-bot-1            |     sys.exit(main())
music-bot-1            |              ^^^^^^
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/memray/commands/__init__.py", line 137, in main
music-bot-1            |     arg_values.entrypoint(arg_values, parser)
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/memray/commands/run.py", line 326, in run
music-bot-1            |     _run_with_file_output(args)
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/memray/commands/run.py", line 175, in _run_with_file_output
music-bot-1            |     _run_tracker(
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/memray/commands/run.py", line 58, in _run_tracker
music-bot-1            |     runpy.run_module(args.script, run_name="__main__", alter_sys=True)
music-bot-1            |   File "/usr/local/lib/python3.11/runpy.py", line 226, in run_module
music-bot-1            |     return _run_module_code(code, init_globals, run_name, mod_spec)
music-bot-1            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
music-bot-1            |   File "/usr/local/lib/python3.11/runpy.py", line 98, in _run_module_code
music-bot-1            |     _run_code(code, mod_globals, init_globals,
music-bot-1            |   File "/usr/local/lib/python3.11/runpy.py", line 88, in _run_code
music-bot-1            |     exec(code, run_globals)
music-bot-1            |   File "/bot/vibr/__main__.py", line 5, in <module>
music-bot-1            |     import uvloop
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/uvloop/__init__.py", line 6, in <module>
music-bot-1            |     from . import includes as __includes  # NOQA
music-bot-1            |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
music-bot-1            |   File "/usr/local/lib/python3.11/site-packages/uvloop/includes/__init__.py", line 16, in <module>
music-bot-1            |     import ssl
music-bot-1            |   File "/usr/local/lib/python3.11/ssl.py", line 100, in <module>
music-bot-1            |     import _ssl             # if we can't import it, let the error propagate
music-bot-1            |     ^^^^^^^^^^^
music-bot-1            | ImportError: /usr/local/lib/python3.11/lib-dynload/_ssl.cpython-311-x86_64-linux-gnu.so: undefined symbol: _PyModule_Add
```

**Reply by godlygeek** [reply]
That error doesn't look like it has anything to do with Memray... what happens if you run:
```
python3.11-dbg -m vibr
```
instead?

**Reply by ooliver1** [reply]
Thank you for sticking with me.

I **do** have yet _another_ issue though, upon exiting I have ``python3.11-dbg: ../Python/compile.c:8160: insert_prefix_instructions: Assertion `c->u->u_firstlineno > 0' failed.`` which is quite  wild

---

Managed to get the undefined symbol error to work with the following (janky) dockerfile:

```dockerfile
FROM python:3.11-slim-bookworm

WORKDIR /bot

RUN apt-get update && apt-get install git netcat-openbsd iproute2 debuginfod python3.11-dbg python3-pip -yqq && rm -rf /var/lib/apt/lists/*

ENV DEBUGINFOD_URLS="https://debuginfod.elfutils.org/"
ENV PYTHONPATH="/usr/lib/python3.11:/usr/lib/python3.11/lib-dynload:/usr/lib/python3.11/site-packages"
RUN python3.11-dbg -m pip install poetry --break-system-packages

COPY vibr/pyproject.toml vibr/poetry.lock ./

RUN poetry install --no-root --only main

COPY . .
RUN mkdir wow

CMD ["sh", "vibr/entrypoint.sh"]
```

- set `PYTHONPATH` to not use `/local/lib/*` but `/lib/*`
- install pip via python3-pip
- use that, with `--break-system-packages` :D
- use a poetry venv which still keeps debug info

**Reply by godlygeek** [reply]
It doesn't seem like you should be touching `PYTHONPATH` at all. I'm not sure if this is the cause of the assertion that you're getting or not, but it could be that you're using forcing modules compiled for one build of the interpreter to be loaded into a different, ABI-incompatible build.

**Comment by ooliver1** [comment]
~~I suppose I might have more luck with not using pre-aggregated, and just hope I have enough storage to hold it, since unless if it has something to do with the stack, there's some random allocation somewhere causing this which is released later.~~

Ignore, reply solved this speculation.

**Reply by godlygeek** [reply]
I don't think this will give you any more information. The flame graph and summary report are both showing you information about the point in time when the process was at its heap memory high water mark, so even large allocations which were later freed will show up.

**Comment by ooliver1** [comment]
@godlygeek I return with a couple of 220M native+aggregated files made via alpine and python3.11 debug.

`memray table memray.368.bin` (also happens with other commands)

```py
Traceback (most recent call last):
  File "/usr/bin/memray", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/lib/python3.11/site-packages/memray/commands/__init__.py", line 137, in main
    arg_values.entrypoint(arg_values, parser)
  File "/usr/lib/python3.11/site-packages/memray/commands/common.py", line 264, in run
    self.write_report(
  File "/usr/lib/python3.11/site-packages/memray/commands/common.py", line 171, in write_report
    reporter = self.reporter_factory(
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/memray/reporters/table.py", line 37, in from_snapshot
    list(record.hybrid_stack_trace(max_stacks=1))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "src/memray/_memray.pyx", line 338, in memray._memray.AllocationRecord.hybrid_stack_trace
  File "src/memray/_memray.pyx", line 245, in memray._memray.hybrid_stack_trace
AssertionError
```

<details>
<summary>Dockerfile</summary>

```dockerfile
FROM alpine

WORKDIR /bot

RUN apk add --no-cache --update git netcat-openbsd iproute2 python3 python3-dbg py3-pip python3-dev gcc
# ENV DEBUGINFOD_URLS="https://debuginfod.elfutils.org/"
RUN python3 -m pip install poetry --break-system-packages

# COPY vibr/pyproject.toml vibr/poetry.lock ./
COPY vibr/requirements.txt ./

RUN pip install -r requirements.txt

COPY . .
RUN mkdir wow

CMD ["sh", "vibr/entrypoint.sh"]
```

</details>

<details>
<summary>entrypoint.sh</summary>

```sh
# unrelated migrations + network waiting

python3 -m memray run --aggregate --native -o wow/memray.$((1 + $RANDOM % 999)).bin -m vibr
```

</details>

<details>
<summary>requirements.txt</summary>

```txt
aiodns==3.0.0 ; python_version >= "3.10" and python_version < "4.0"
aiohttp==3.8.4 ; python_version >= "3.10" and python_version < "4.0"
aiohttp[speedups]==3.8.4 ; python_version >= "3.10" and python_version < "4.0"
aiosignal==1.3.1 ; python_version >= "3.10" and python_version < "4.0"
async-spotify @ git+https://github.com/ooliver1/AsyncSpotify@dc09920#subdirectory=src ; python_version >= "3.10" and python_version < "4.0"
async-timeout==4.0.2 ; python_version >= "3.10" and python_version < "4.0"
asyncache==0.3.1 ; python_version >= "3.10" and python_version < "4.0"
asyncpg==0.27.0 ; python_version >= "3.10" and python_version < "4.0"
attrs==23.1.0 ; python_version >= "3.10" and python_version < "4.0"
black==23.3.0 ; python_version >= "3.10" and python_version < "4.0"
brotli==1.0.9 ; python_version >= "3.10" and python_version < "4.0"
cachetools==5.3.0 ; python_version >= "3.10" and python_version < "4.0"
certifi==2023.5.7 ; python_version >= "3.10" and python_version < "4.0"
cffi==1.15.1 ; python_version >= "3.10" and python_version < "4.0"
charset-normalizer==3.1.0 ; python_version >= "3.10" and python_version < "4.0"
click==8.1.3 ; python_version >= "3.10" and python_version < "4.0"
colorama==0.4.6 ; python_version >= "3.10" and python_version < "4.0"
delarva @ git+https://github.com/ooliver1/delarva@0391e39 ; python_version >= "3.10" and python_version < "4.0"
dnspython==2.3.0 ; python_version >= "3.10" and python_version < "4.0"
docker==6.1.2 ; python_version >= "3.10" and python_version < "4.0"
docstring-parser==0.15 ; python_version >= "3.10" and python_version < "4.0"
email-validator==2.0.0.post2 ; python_version >= "3.10" and python_version < "4.0"
frozenlist==1.3.3 ; python_version >= "3.10" and python_version < "4.0"
hiredis==2.2.3 ; python_version >= "3.10" and python_version < "4.0"
humanfriendly==10.0 ; python_version >= "3.10" and python_version < "4.0"
idna==3.4 ; python_version >= "3.10" and python_version < "4.0"
inflection==0.5.1 ; python_version >= "3.10" and python_version < "4.0"
jinja2==3.1.2 ; python_version >= "3.10" and python_version < "4.0"
mafic @ git+https://github.com/ooliver1/mafic@master ; python_version >= "3.10" and python_version < "4.0"
markdown-it-py==3.0.0 ; python_version >= "3.10" and python_version < "4.0"
markupsafe==2.1.2 ; python_version >= "3.10" and python_version < "4.0"
mdurl==0.1.2 ; python_version >= "3.10" and python_version < "4.0"
memray==1.9.1 ; python_version >= "3.10" and python_version < "4.0"
mslex==0.3.0 ; python_version >= "3.10" and python_version < "4.0" and sys_platform == "win32"
multidict==6.0.4 ; python_version >= "3.10" and python_version < "4.0"
mypy-extensions==1.0.0 ; python_version >= "3.10" and python_version < "4.0"
nextcord @ git+https://github.com/boting-devs/nextcord@vibr ; python_version >= "3.10" and python_version < "4.0"
nextcord-ext-menus==1.5.6 ; python_version >= "3.10" and python_version < "4.0"
nextcord[speed] @ git+https://github.com/boting-devs/nextcord@vibr ; python_version >= "3.10" and python_version < "4.0"
ooliver-botbase[db] @ git+https://github.com/ooliver1/botbase@no-psutil ; python_version >= "3.10" and python_version < "4.0"
orjson==3.8.14 ; python_version >= "3.10" and python_version < "4.0"
packaging==23.1 ; python_version >= "3.10" and python_version < "4.0"
pathspec==0.11.1 ; python_version >= "3.10" and python_version < "4.0"
piccolo[orjson,postgres,uvloop]==0.111.1 ; python_version >= "3.10" and python_version < "4.0"
platformdirs==3.5.1 ; python_version >= "3.10" and python_version < "4.0"
prometheus-async==22.2.0 ; python_version >= "3.10" and python_version < "4.0"
prometheus-client==0.17.0 ; python_version >= "3.10" and python_version < "4.0"
pycares==4.3.0 ; python_version >= "3.10" and python_version < "4.0"
pycparser==2.21 ; python_version >= "3.10" and python_version < "4.0"
pydantic[email]==1.10.8 ; python_version >= "3.10" and python_version < "4.0"
pygments==2.16.1 ; python_version >= "3.10" and python_version < "4.0"
pyreadline3==3.4.1 ; sys_platform == "win32" and python_version >= "3.10" and python_version < "4.0"
python-dotenv==1.0.0 ; python_version >= "3.10" and python_version < "4.0"
pywin32==306 ; python_version >= "3.10" and python_version < "4.0" and sys_platform == "win32"
pyyaml==6.0.1 ; python_version >= "3.10" and python_version < "4.0"
redis[hiredis]==4.5.5 ; python_version >= "3.10" and python_version < "4.0"
requests==2.31.0 ; python_version >= "3.10" and python_version < "4.0"
rich==13.5.2 ; python_version >= "3.10" and python_version < "4.0"
targ==0.3.8 ; python_version >= "3.10" and python_version < "4.0"
tomli==2.0.1 ; python_version >= "3.10" and python_version < "4.0"
typing-extensions==4.6.1 ; python_version >= "3.10" and python_version < "4.0"
urllib3==2.0.2 ; python_version >= "3.10" and python_version < "4.0"
uvloop==0.17.0 ; python_version >= "3.10" and python_version < "4.0"
websocket-client==1.5.2 ; python_version >= "3.10" and python_version < "4.0"
wrapt==1.15.0 ; python_version >= "3.10" and python_version < "4.0"
yarl==1.9.2 ; python_version >= "3.10" and python_version < "4.0"
```

</details>

[memray.zip](https://ufile.io/9m9fae86)
Sorry for the site, too big to fit here.

**Reply by pablogsal** [reply]
Hi @ooliver1, I tried to reproduce but unfortunately I am getting

```
ImportError: No module named vibr
```

can you provide the link to the `vibr` repo or how to install that in the Docker container?

**Reply by ooliver1** [reply]
@pablogsal It is closed source, I cannot. I was more really wondering how I can fix my existing memray files, since I really can't run this again since it takes 4 days of semi-downtime to get it to show up (large CPU spikes at peak for native tracking) as rate of `process_cpu_seconds_total`.

![image](https://github.com/bloomberg/memray/assets/34910574/b56b23bb-d047-4dab-adca-503eb84fdbf5)

And a refresh of the memory 'leak'(?)'s metrics in GiB of resident memory:

![image](https://github.com/bloomberg/memray/assets/34910574/4ca5163d-5ffd-4fcc-9734-545e59faeb5c)

**Reply by pablogsal** [reply]
I have analyzed the files you provide and seems that there is only one record with some corrupt stack. Is unclear how that happens (for that we would need a reproducer on how to generate the file). But for the time being, you can patch memray locally to investigate your problem. To do that surround these lines with a try-except:

https://github.com/bloomberg/memray/blob/214debfcb47f3e1dba555c665f6643f263f18a6b/src/memray/reporters/flamegraph.py#L128-L132

In this way:


```
try:
        stack = (
                tuple(record.hybrid_stack_trace())
                if native_traces
                else record.stack_trace()
            )
except Exception:
    print(f"Ignoring record {record}")
    continue
```

You can modify the files locally by checking where memray is installed:

```
>>> import memray.reporters.flamegraph as f
>>> f
<module 'memray.reporters.flamegraph' from '/path/to/memray/reporters/flamegraph.py'>
```

From this, you know that you need to edit the file ` '/path/to/memray/reporters/flamegraph.py'`. We will probably do a release with some fix aking to this one but for the time being this will allow you to continue your investigation

**Reply by pablogsal** [reply]
Btw, I noticed that malloc has been returned NULL a ton of times. There are a lot of refords with address `0x0` in the file:

```
AllocationRecord<tid=0x1, address=0x0, size=2.000MB, allocator=10, allocations=2> 
```

@godlygeek I think we probably should not count those or we should warn or something because that probably is not counting towards maximum heap.

**Reply by pablogsal** [reply]
The other problem is that there is a LOT of memory records and the flamegraph cannot display that many points. You can reduce this by appliying this patch:

```diff
--- a/src/memray/reporters/flamegraph.py
+++ b/src/memray/reporters/flamegraph.py
@@ -118,7 +118,7 @@ class FlameGraphReporter:
     ) -> None:
         super().__init__()
         self.data = data
-        self.memory_records = memory_records
+        self.memory_records = memory_records[::1000]

     @classmethod
```

**Reply by pablogsal** [reply]
Notice that to render a native flamegraph you need to run the command inside the docker container or otherwise you won't resolve symbols.

**Reply by ooliver1** [reply]
@pablogsal I appareciate the in-depth replies. I've been running it in the same docker container as before, with a few extra libraries installed to get memray to compile. I have [lots of ignoring record logs](https://workbin.dev/?id=1694370956120175), [some missing shared objects](https://workbin.dev/?id=1694371059351921) (using alpine as the development build clashes with debian), and the flamegraph only shows the expected memory usage (using both `::1000` and `:1000` - was unsure if `::1000` was intentional but that doesn't matter).

![image](https://github.com/bloomberg/memray/assets/34910574/25674687-ef99-4c88-8cb1-22b5fad2c850)
![image](https://github.com/bloomberg/memray/assets/34910574/71b2228a-553a-4834-b512-c4ca457d6b41)

How I've been running this container locally is building it and assigning the local path to a volume within docker, and simply running `ash` as the command when running the container. That way all you need is the requirements.txt.

**Reply by pablogsal** [reply]
Not sure if I understand correctly the conclusion to your last comment but here are some observations:

* You need to run memray and "memray flamegraph" on the same container, otherwise you are going to get a lot of missing libraries. I would recommend to modify your entry point script to also run memray flamegraph or instead use "-it" to enter an interactive docker shell and run it yourself.

* You don't need to recompile memray. Just run an interactive shell, install vim or nano in the container and modify the python file as I mentioned before.

* In any case, after analysing your result file I am very perplexed. There are a ton of records where the allocator just returns NULL (the majority) and the python stacks doesn't make any sense to me. Something weird is going on there but unfortunately without access to a complete reproducer we cannot do anything else on our side.

**Reply by ooliver1** [reply]
1. It is in the same container (well, image, the container has to exit before the file is created)
2. Compile as in my fork containing your mentioned changes
3. I could send you a zip of the project and some development env vars (still 'sensitive' but not damaging to send to a trusted source. Where could I send you it?

**Reply by pablogsal** [reply]
> 1. It is in the same container (well, image, the container has to exit before the file is created)
> 
> 2. Compile as in my fork containing your mentioned changes
> 
> 3. I could send you a zip of the project and some development env vars (still 'sensitive' but not damaging to send to a trusted source. Where could I send you it?

Instead of sending the project, try to trim your program as much as possible while showing the same problem and then try to produce a minimal self contained reproducer. Even if you send us the entire project we are unlikely to be able to help if is too big or requires a lot of work to reproduce. You don't need to produce the leak, just something that crashes memray when checking the results with that assertion.

I know it sounds complicated but unfortunately we won't be able to do a lot without something like that we can investigate in a reasonable amount of time.

Maybe you can start running memray with a request or two and checking if that works.

**Reply by ooliver1** [reply]
I can't even get it to happen on my machine when running locally, it is only something that happens with a ton of users using it constantly, so I suppose I'm stuck for now. Thank you for helping though!

It may happen if I run it for a few hours, but I can't really do that right now.

Update: I have a couple of ideas for it and have managed to reproduce it locally with enough usage, thankfully I can finally start to strip away at the project, and I will hopefully be back soon.
